File: Data Structures & Algorithms/cheapest-flight-path/submission-4.java
class Solution {
    Map<Integer, List<int[]>> m = new HashMap<>();
    int dst;
    Integer[][] memo;
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        this.dst = dst;
        memo = new Integer[n][k + 2];        // stopsLeft 范围 0..k+1
        for (int[] f : flights) {
            m.computeIfAbsent(f[0], key -> new ArrayList<>()).add(new int[]{f[1],f[2]});
        }
        int ans = dfs(src, k+1);
        return ans == Integer.MAX_VALUE ? -1 : ans;
    }
    // 从 cur 到 dst、最多再走 stopsLeft 条边的最小花费
    int dfs(int cur, int stopsLeft) {
        if (cur == dst) return 0;
        if (stopsLeft == 0) return Integer.MAX_VALUE;
        if (memo[cur][stopsLeft] != null) return memo[cur][stopsLeft];
        int best = Integer.MAX_VALUE;
        for (int[] next : m.getOrDefault(cur, new ArrayList<>())) {
            int sub = dfs(next[0], stopsLeft - 1);
            if (sub != Integer.MAX_VALUE) best = Math.min(best, sub + next[1]);
        }
        return memo[cur][stopsLeft] = best;
    }
}

File: Data Structures & Algorithms/longest-consecutive-sequence/submission-4.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        s = set()
        for n in nums:
            s.add(n)
        
        ans = 0
        for n in nums:
            if n-1 in s:
                continue
            cur = n
            cur_len = 1
            while (cur+1) in s:
                cur_len += 1
                cur+= 1
            ans = max(ans, cur_len)
        
        return ans